sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/HttpResponse.java

```diff
@@ -0,0 +1,165 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kalix.javasdk;
+
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public interface HttpResponse {
+  StatusCode.Success getStatusCode();
+
+  Optional<String> getContentType();
+
+  byte[] getBody();
+
+  static HttpResponse ok() {
+    return new NoContentResponse(StatusCode.Success.OK);
+  }
+
+  static HttpResponse created() {
+    return new NoContentResponse(StatusCode.Success.CREATED);
+  }
+
+  static HttpResponse empty(StatusCode.Success statusCode) {
+    return new NoContentResponse(statusCode);
+  }
+
+  static ByteContentResponse text(String text) {
+    return new ByteContentResponse(StatusCode.Success.OK, "text/plain", text.getBytes(UTF_8));
+  }
+
+  static JsonContentResponse json(Object object) {
+    return new JsonContentResponse(StatusCode.Success.OK, object);
+  }
+
+  static JsonContentResponse json(StatusCode.Success statusCode, Object object) {
+    return new JsonContentResponse(statusCode, object);
+  }
+
+  static ByteContentResponse bytes(byte[] body) {
+    return new ByteContentResponse(StatusCode.Success.OK, "application/octet-stream", body);
+  }
+
+  class NoContentResponse implements HttpResponse {
+
+    private final StatusCode.Success statusCode;
+
+    private NoContentResponse(StatusCode.Success statusCode) {
+      this.statusCode = statusCode;
+    }
+
+    @Override
+    public StatusCode.Success getStatusCode() {
+      return statusCode;
+    }
+
+    @Override
+    public Optional<String> getContentType() {
+      return Optional.empty();
+    }
+
+    @Override
+    public byte[] getBody() {
+      return new byte[0];
+    }
+  }
+
+  class JsonContentResponse implements HttpResponse {
+
+    private final StatusCode.Success statusCode;
+    private final Object object;
+    private final byte[] body;
+
+    JsonContentResponse(StatusCode.Success statusCode, Object object) {
+      this.statusCode = statusCode;
+      this.object = object;
+      try {
+        this.body = JsonSupport.encodeToBytes(object).toByteArray();
+      } catch (JsonProcessingException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public StatusCode.Success getStatusCode() {
+      return statusCode;
+    }
+
+    @Override
+    public Optional<String> getContentType() {
+      return Optional.of("application/json");
+    }
+
+    public Object getObject() {
+      return object;
+    }
+
+    @Override
+    public byte[] getBody() {
+      return body;
+    }
+
+    public <T> T bodyAs(Class<T> clz) throws IOException {
+      return JsonSupport.parseBytes(body, clz);
+    }
```

Comment:
I realise now that this method can only be useful if added to the HttpResponse. 

When we receive a response from the ComponentClient, we will take the internal http response from from the WebClient and map it to HttpResponse, no? 

Actually, it seems to me that the only public type should be `HttpResponse`. 

So, in that case, it must communicate in its name that it is only applicable with payload is json. Something like `jsonBodyAs(Class<T> clz)`

Reply:
Having `HttpResponse` as the only public class would introduce another problems, for instance we allow to change the content-type only in case of `ByteContentResponse`. If we move this method to the interface, then other implementations will throw an exception. But I agree that in fact you will not be able to use this method, because I'm parsing the response always to `ByteContentReponse`.

Reply:
Maybe it's time to move one step back and remove all the specific implementations and have only a single class with factory methods? and then we could have `jsonBodyAs(Class<T> clz)`?

Reply:
Yeah, I was having the same thought. We can keep the specific methods, but one could always change the content-type and status code later. Not a good idea, but it will make things much simpler. 

So, for example, if I really want to mess-up, I can do `HttpResponse.ok(Foo)`. That will return a 200 with json. Then `HttpResponse.ok(Foo).withContentType(whatever).withStatusCode(NotFound)`. 

I would be messing up, but then we can argue that the user chose to do so. 🤷

Reply:
An alternative could be that we only offer a method to change the content-type, but not the status code nor the body. This is determined by the helper method `ok(T)`, `ok(byte[])`, `noContent()`, `notFound()`, etc.

Reply:
I would have more faith in our users. This is low-level api, they shouldn't use it for normal cases, we just want to support very specific cases.

Reply:
You can check the implementation now.